contracts/core/AuxiliaryBlockStore.sol

```diff
@@ -0,0 +1,214 @@
+            accumulatedGases[header.blockHash] = accumulatedGases[header.parentHash].add(
+                header.gasUsed
+            );
+
+            accumulatedTransactionRoots[header.blockHash] = keccak256(
+                abi.encode(
+                    accumulatedTransactionRoots[header.parentHash],
+                    header.transactionRoot
+                )
+            );
+        }
+    }
+
+    /* Internal Functions */
+
+    /**
+     * @notice Checks if a target block is valid. The same criteria apply for
+     *         voting and justifying, as justifying results from voting.
+     *
+     * @dev In addition to the general block store's check whether a target is
+     *      valid, here it also has to be a descendant of the source block.
+     *
+     * @param _sourceBlockHash The hash of the corresponding source.
+     * @param _targetBlockHash The hash of the potential target block.
+     *
+     * @return valid_ `true` if the given block hash is a valid target.
+     * @return reason_ Gives the reason in case the block is not a valid target.
+     */
+    function isTargetValid(
+        bytes32 _sourceBlockHash,
+        bytes32 _targetBlockHash
+    )
+        internal
+        view
+        returns (bool valid_, string reason_)
+    {
+        (valid_, reason_) = super.isTargetValid(
+            _sourceBlockHash,
+            _targetBlockHash
+        );
+
+        if (
+            valid_ &&
+            !isAncestorCheckpoint(_sourceBlockHash, _targetBlockHash)
+        ) {
+            valid_ = false;
+            reason_ = "The source must be an ancestor of the target.";
+        }
+    }
+
+    /* Private Functions */
+
+    /**
+     * @notice Checks whether there is a chain of blocks and checkpoints from
+     *         the ancestor to the descendant. It first goes back from the
+     *         descendant along blocks to the latest checkpoint. Then it
+     *         continues along the list of checkpoints and checks whether it can
+     *         reach the ancestor.
+     *
+     * @param _ancestor The block hash of the ancestor. Must be a justified
+     *                  checkpoint.
+     * @param _descendant The block hash of the descendant. Mist be a reported
+     *                    block.
+     *
+     * @return isAncestor_ `true` if it finds a path from the descendant to the
+     *                     ancestor.
+     */
+    function isAncestorCheckpoint(
+        bytes32 _ancestor,
+        bytes32 _descendant
+    )
+        private
+        view
+        returns (bool isAncestor_)
+    {
+        /*
+         * Walk along list of blocks to closest checkpoint. The descendant is
+         * not (yet) a justified checkpoint, so we first need to find the
+         * closest (youngest) checkpoint.
+         */
+        Block.Header storage currentBlock = reportedBlocks[_descendant];
+        while (
+            checkpoints[currentBlock.blockHash].blockHash != currentBlock.blockHash
```

Comment:
It is only equal if the checkpoint is registered/in storage. Otherwise, ig the checkpoint does not exist, `checkpoints[currentBlock.blockHash]` will point to an empty storage location and `checkpoints[currentBlock.blockHash].blockHash` will be the default value `0x00...`.